contracts/access/TimelockController.sol

```diff
@@ -0,0 +1,294 @@
+    function _schedule(bytes32 id, uint256 delay) internal {
+        require(_timestamps[id] == 0, "TimelockController: operation already scheduled");
+        require(delay >= _minDelay, "TimelockController: insufficient delay");
+        // solhint-disable-next-line not-rely-on-time
+        _timestamps[id] = SafeMath.add(block.timestamp, delay);
+    }
+
+    /**
+     * @dev Cancel an operation.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'proposer' role.
+     */
+    function cancel(bytes32 id) public onlyRole(PROPOSER_ROLE) {
+        require(!isOperationDone(id), "TimelockController: operation is already executed");
+        delete _timestamps[id];
+
+        emit Cancelled(id);
+    }
+
+    /**
+     * @dev Execute an (ready) operation containing a single transaction.
+     *
+     * Emits a {CallExecuted} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'executor' role.
+     */
+    function execute(address target, uint256 value, bytes calldata data, bytes32 predecessor, bytes32 salt) public payable onlyRole(EXECUTOR_ROLE) {
+        bytes32 id = hashOperation(target, value, data, predecessor, salt);
+        _execute(id, predecessor);
+        _call(id, 0, target, value, data);
+    }
+
+    /**
+     * @dev Execute an (ready) operation containing a batch of transactions.
+     *
+     * Emits one {CallExecuted} event per transaction in the batch.
+     *
+     * Requirements:
+     *
+     * - the caller must have the 'executor' role.
+     */
+    function executeBatch(address[] calldata targets, uint256[] calldata values, bytes[] calldata datas, bytes32 predecessor, bytes32 salt) public payable onlyRole(EXECUTOR_ROLE) {
+        require(targets.length == values.length, "TimelockController: length missmatch");
+        require(targets.length == datas.length, "TimelockController: length missmatch");
+
+        bytes32 id = hashOperationBatch(targets, values, datas, predecessor, salt);
+        _execute(id, predecessor);
+        for (uint256 i = 0; i < targets.length; ++i) {
+            _call(id, i, targets[i], values[i], datas[i]);
+        }
+    }
+
+    /**
+     * @dev Execute an operation. Operation must be ready.
+     */
+    function _execute(bytes32 id, bytes32 predecessor) internal {
+        require(isOperationReady(id), "TimelockController: operation is not ready");
+        require(predecessor == bytes32(0) || isOperationDone(predecessor), "TimelockController: missing dependency");
+        _timestamps[id] = _DONE_TIMESTAMP;
+    }
+
+    /**
+     * @dev Execute a transaction.
+     *
+     * Emits a {CallExecuted} event.
+     */
+    function _call(bytes32 id, uint256 index, address target, uint256 value, bytes calldata data) internal returns (bool) {
+        // solhint-disable-next-line avoid-low-level-calls
+        (bool success,) = target.call{value: value}(data);
+        require(success, "TimelockController: underlying transaction reverted");
```

Comment:
I would like to use [`Address.functionCall`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/09d437c1f9f68c4c242db8e489cf3ab882b7a1f3/contracts/utils/Address.sol#L79) here, which handles the revert nicely by forwarding a revert reason if there is one. However, it will revert if the target is not a contract, so it wouldn't work for simple value transfers.

Reply:
I think the ability to send value to EOA is essential, has the timelock might receive ETH as part of maintenance operations. (some contract might send value to the owner ... of might send funds back to the owner in case of selfdestruct...)

So yes for `Address.functionCall` unless it reverts when calling EOAs

Reply:
Yes, I agree. I edited my comment after your reply but before seeing it. My conclusion:

>Maybe we can leave this to be improved by a subsequent PR.